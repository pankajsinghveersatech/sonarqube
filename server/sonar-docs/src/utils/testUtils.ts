/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
