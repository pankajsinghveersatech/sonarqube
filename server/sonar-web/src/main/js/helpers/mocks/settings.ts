/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Setting, SettingCategoryDefinition, SettingWithCategory } from '../../types/settings';

export function mockDefinition(
  overrides: Partial<SettingCategoryDefinition> = {}
): SettingCategoryDefinition {
  return {
    key: 'foo',
    category: 'foo category',
    fields: [],
    options: [],
    subCategory: 'foo subCat',
    ...overrides
  };
}

export function mockSetting(overrides: Partial<Setting> = {}): Setting {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: []
    },
    ...overrides
  };
}

export function mockSettingWithCategory(
  overrides: Partial<SettingWithCategory> = {}
): SettingWithCategory {
  return {
    key: 'foo',
    value: '42',
    inherited: true,
    definition: {
      key: 'foo',
      name: 'Foo setting',
      description: 'When Foo then Bar',
      type: 'INTEGER',
      options: [],
      category: 'general',
      fields: [],
      subCategory: 'email'
    },
    ...overrides
  };
}
