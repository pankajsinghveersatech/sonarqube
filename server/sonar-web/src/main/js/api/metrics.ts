/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import throwGlobalError from '../app/utils/throwGlobalError';
import { getJSON } from '../helpers/request';

export interface MetricsResponse {
  metrics: T.Metric[];
  p: number;
  ps: number;
  total: number;
}

export function getMetrics(data?: {
  isCustom?: boolean;
  p?: number;
  ps?: number;
}): Promise<MetricsResponse> {
  return getJSON('/api/metrics/search', data).catch(throwGlobalError);
}

export function getAllMetrics(data?: {
  isCustom?: boolean;
  p?: number;
  ps?: number;
}): Promise<T.Metric[]> {
  return inner(data);

  function inner(
    data: { p?: number; ps?: number } = { ps: 500 },
    prev?: MetricsResponse
  ): Promise<T.Metric[]> {
    return getMetrics(data).then(r => {
      const result = prev ? prev.metrics.concat(r.metrics) : r.metrics;
      if (r.p * r.ps >= r.total) {
        return result;
      }
      return inner({ ...data, p: r.p + 1 }, { ...r, metrics: result });
    });
  }
}
