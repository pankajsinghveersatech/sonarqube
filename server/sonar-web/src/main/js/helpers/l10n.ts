/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { fetchL10nBundle } from '../api/l10n';
import { L10nBundle, L10nBundleRequestParams } from '../types/l10n';
import { toNotSoISOString } from './dates';
import { get as loadFromLocalStorage, save as saveInLocalStorage } from './storage';

export type Messages = T.Dict<string>;

export const DEFAULT_LOCALE = 'en';
export const DEFAULT_MESSAGES = {
  // eslint-disable-next-line camelcase
  default_error_message: 'The request cannot be processed. Try again later.'
};

let allMessages: Messages = {};
let locale: string | undefined;

export function translate(...keys: string[]): string {
  const messageKey = keys.join('.');
  const l10nMessages = getMessages();
  if (process.env.NODE_ENV === 'development' && !l10nMessages[messageKey]) {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return l10nMessages[messageKey] || messageKey;
}

export function translateWithParameters(
  messageKey: string,
  ...parameters: Array<string | number>
): string {
  const message = getMessages()[messageKey];
  if (message) {
    return parameters
      .map(parameter => String(parameter))
      .reduce((acc, parameter, index) => acc.replace(`{${index}}`, () => parameter), message);
  }
  if (process.env.NODE_ENV === 'development') {
    // eslint-disable-next-line no-console
    console.error(`No message for: ${messageKey}`);
  }
  return `${messageKey}.${parameters.join('.')}`;
}

export function hasMessage(...keys: string[]): boolean {
  const messageKey = keys.join('.');
  return getMessages()[messageKey] != null;
}

export function getMessages() {
  if (typeof allMessages === 'undefined') {
    logWarning('L10n messages are not initialized. Use default messages.');
    return DEFAULT_MESSAGES;
  }
  return allMessages;
}

export function resetMessages(newMessages: Messages) {
  allMessages = newMessages;
}

export function getLocalizedMetricName(
  metric: { key: string; name?: string },
  short = false
): string {
  const bundleKey = `metric.${metric.key}.${short ? 'short_name' : 'name'}`;
  if (hasMessage(bundleKey)) {
    return translate(bundleKey);
  } else if (short) {
    return getLocalizedMetricName(metric);
  }
  return metric.name || metric.key;
}

export function getLocalizedCategoryMetricName(metric: { key: string; name?: string }) {
  const bundleKey = `metric.${metric.key}.extra_short_name`;
  return hasMessage(bundleKey) ? translate(bundleKey) : getLocalizedMetricName(metric, true);
}

export function getLocalizedMetricDomain(domainName: string) {
  const bundleKey = `metric_domain.${domainName}`;
  return hasMessage(bundleKey) ? translate(bundleKey) : domainName;
}

export function getCurrentLocale() {
  return locale;
}

export function resetCurrentLocale(newLocale: string) {
  locale = newLocale;
}

export function getShortMonthName(index: number) {
  const months = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec'
  ];
  return translate(months[index]);
}

export function getWeekDayName(index: number) {
  const weekdays = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

export function getShortWeekDayName(index: number) {
  const weekdays = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  return weekdays[index] ? translate(weekdays[index]) : '';
}

const L10N_BUNDLE_LS_KEY = 'l10n.bundle';

export async function loadL10nBundle() {
  const bundle = await getLatestL10nBundle().catch(() => ({
    locale: DEFAULT_LOCALE,
    messages: {}
  }));

  resetCurrentLocale(bundle.locale);
  resetMessages(bundle.messages);

  // No need to load english (default) bundle, it's coming with react-intl
  if (bundle.locale !== DEFAULT_LOCALE) {
    const [intlBundle, intl] = await Promise.all([
      import(`react-intl/locale-data/${bundle.locale}`),
      import('react-intl')
    ]);

    intl.addLocaleData(intlBundle.default);
  }

  return bundle;
}

export async function getLatestL10nBundle() {
  const browserLocale = getPreferredLanguage();
  const cachedBundle = loadL10nBundleFromLocalStorage();

  const params: L10nBundleRequestParams = {};

  if (browserLocale) {
    params.locale = browserLocale;

    if (
      cachedBundle.locale &&
      browserLocale.startsWith(cachedBundle.locale) &&
      cachedBundle.timestamp &&
      cachedBundle.messages
    ) {
      params.ts = cachedBundle.timestamp;
    }
  }

  const { effectiveLocale, messages } = await fetchL10nBundle(params).catch(response => {
    if (response && response.status === 304) {
      return {
        effectiveLocale: cachedBundle.locale || browserLocale || DEFAULT_LOCALE,
        messages: cachedBundle.messages ?? {}
      };
    }
    throw new Error(`Unexpected status code: ${response.status}`);
  });

  const bundle = {
    timestamp: toNotSoISOString(new Date()),
    locale: effectiveLocale,
    messages
  };

  saveL10nBundleToLocalStorage(bundle);

  return bundle;
}

export function getCurrentL10nBundle() {
  return loadL10nBundleFromLocalStorage();
}

function getPreferredLanguage() {
  return window.navigator.languages ? window.navigator.languages[0] : window.navigator.language;
}

function loadL10nBundleFromLocalStorage() {
  let bundle: L10nBundle;

  try {
    bundle = JSON.parse(loadFromLocalStorage(L10N_BUNDLE_LS_KEY) ?? '{}');
  } catch {
    bundle = {};
  }

  return bundle;
}

function saveL10nBundleToLocalStorage(bundle: L10nBundle) {
  saveInLocalStorage(L10N_BUNDLE_LS_KEY, JSON.stringify(bundle));
}

function logWarning(message: string) {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.warn(message);
  }
}
