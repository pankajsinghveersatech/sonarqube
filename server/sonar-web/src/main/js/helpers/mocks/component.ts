/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier, TreeComponent } from '../../types/component';
import { MetricKey } from '../../types/metrics';
import { mockMeasureEnhanced } from '../testMocks';

export function mockComponent(overrides: Partial<T.Component> = {}): T.Component {
  return {
    breadcrumbs: [],
    key: 'my-project',
    name: 'MyProject',
    qualifier: ComponentQualifier.Project,
    qualityGate: { isDefault: true, key: '30', name: 'Sonar way' },
    qualityProfiles: [
      {
        deleted: false,
        key: 'my-qp',
        language: 'ts',
        name: 'Sonar way'
      }
    ],
    tags: [],
    ...overrides
  };
}

export function mockTreeComponent(overrides: Partial<TreeComponent>): TreeComponent {
  return {
    key: 'my-key',
    qualifier: ComponentQualifier.Project,
    name: 'component',
    visibility: 'public',
    ...overrides
  };
}

export function mockComponentMeasure(
  file = false,
  overrides: Partial<T.ComponentMeasure> = {}
): T.ComponentMeasure {
  if (file) {
    return {
      key: 'foo:src/index.tsx',
      name: 'index.tsx',
      qualifier: ComponentQualifier.File,
      path: 'src/index.tsx',
      measures: [{ metric: MetricKey.bugs, value: '1', bestValue: false }],
      ...overrides
    };
  }
  return {
    key: 'foo',
    name: 'Foo',
    qualifier: ComponentQualifier.Project,
    measures: [{ metric: MetricKey.bugs, value: '12', bestValue: false }],
    ...overrides
  };
}

export function mockComponentMeasureEnhanced(
  overrides: Partial<T.ComponentMeasureEnhanced> = {}
): T.ComponentMeasureEnhanced {
  return {
    ...mockComponentMeasure(false, overrides as T.ComponentMeasure),
    leak: undefined,
    measures: [mockMeasureEnhanced()],
    value: undefined,
    ...overrides
  };
}
