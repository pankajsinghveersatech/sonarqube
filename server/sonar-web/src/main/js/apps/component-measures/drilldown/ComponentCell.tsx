/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { LocationDescriptor } from 'history';
import * as React from 'react';
import { Link } from 'react-router';
import BranchIcon from '../../../components/icons/BranchIcon';
import LinkIcon from '../../../components/icons/LinkIcon';
import QualifierIcon from '../../../components/icons/QualifierIcon';
import { translate } from '../../../helpers/l10n';
import { splitPath } from '../../../helpers/path';
import {
  getBranchLikeUrl,
  getComponentDrilldownUrlWithSelection,
  getProjectUrl
} from '../../../helpers/urls';
import { BranchLike } from '../../../types/branch-like';
import {
  ComponentQualifier,
  isApplication,
  isPortfolioLike,
  isProject
} from '../../../types/component';
import { MeasurePageView } from '../../../types/measures';
import { MetricKey } from '../../../types/metrics';

export interface ComponentCellProps {
  branchLike?: BranchLike;
  component: T.ComponentMeasureEnhanced;
  metric: T.Metric;
  rootComponent: T.ComponentMeasure;
  view: MeasurePageView;
}

export default function ComponentCell(props: ComponentCellProps) {
  const { branchLike, component, metric, rootComponent, view } = props;

  let head = '';
  let tail = component.name;

  if (
    view === 'list' &&
    ([
      ComponentQualifier.File,
      ComponentQualifier.TestFile,
      ComponentQualifier.Directory
    ] as string[]).includes(component.qualifier) &&
    component.path
  ) {
    ({ head, tail } = splitPath(component.path));
  }

  let path: LocationDescriptor;
  if (component.refKey) {
    if (
      !isPortfolioLike(component.qualifier) &&
      ([MetricKey.releasability_rating, MetricKey.alert_status] as string[]).includes(metric.key)
    ) {
      path = isApplication(component.qualifier)
        ? getProjectUrl(component.refKey, component.branch)
        : getBranchLikeUrl(component.refKey, branchLike);
    } else if (isProject(component.qualifier) && metric.key === MetricKey.projects) {
      path = getBranchLikeUrl(component.refKey, branchLike);
    } else {
      path = getComponentDrilldownUrlWithSelection(
        component.refKey,
        '',
        metric.key,
        branchLike,
        view
      );
    }
  } else {
    path = getComponentDrilldownUrlWithSelection(
      rootComponent.key,
      component.key,
      metric.key,
      branchLike,
      view
    );
  }

  return (
    <td className="measure-details-component-cell">
      <div className="text-ellipsis">
        <Link
          className="link-no-underline"
          to={path}
          id={'component-measures-component-link-' + component.key}>
          {component.refKey && (
            <span className="big-spacer-right">
              <LinkIcon />
            </span>
          )}
          <span title={component.key}>
            <QualifierIcon className="little-spacer-right" qualifier={component.qualifier} />
            {head.length > 0 && <span className="note">{head}/</span>}
            <span>{tail}</span>
            {isApplication(rootComponent.qualifier) &&
              (component.branch ? (
                <>
                  <BranchIcon className="spacer-left little-spacer-right" />
                  <span className="note">{component.branch}</span>
                </>
              ) : (
                <span className="spacer-left badge">{translate('branches.main_branch')}</span>
              ))}
          </span>
        </Link>
      </div>
    </td>
  );
}
