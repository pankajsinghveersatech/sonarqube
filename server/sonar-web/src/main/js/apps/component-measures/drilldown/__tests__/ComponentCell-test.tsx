/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { mockBranch, mockPullRequest } from '../../../../helpers/mocks/branch-like';
import {
  mockComponentMeasure,
  mockComponentMeasureEnhanced
} from '../../../../helpers/mocks/component';
import { mockMetric } from '../../../../helpers/testMocks';
import { ComponentQualifier } from '../../../../types/component';
import { MetricKey } from '../../../../types/metrics';
import { enhanceComponent } from '../../utils';
import ComponentCell, { ComponentCellProps } from '../ComponentCell';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(
    shallowRender({
      rootComponent: mockComponentMeasure(false, { qualifier: ComponentQualifier.Application })
    })
  ).toMatchSnapshot('root component is application, component is on main branch');
  expect(
    shallowRender({
      rootComponent: mockComponentMeasure(false, { qualifier: ComponentQualifier.Application }),
      component: mockComponentMeasureEnhanced({ branch: 'develop' })
    })
  ).toMatchSnapshot('root component is application, component has branch');
  expect(
    shallowRender({ component: mockComponentMeasureEnhanced({ refKey: 'project-key' }) })
  ).toMatchSnapshot('ref project component');
  expect(
    shallowRender(
      {
        component: mockComponentMeasureEnhanced({
          refKey: 'project-key',
          qualifier: ComponentQualifier.Project
        }),
        branchLike: mockBranch()
      },
      MetricKey.releasability_rating
    )
  ).toMatchSnapshot('ref project component, releasability metric');
  expect(
    shallowRender(
      {
        component: mockComponentMeasureEnhanced({
          refKey: 'app-key',
          qualifier: ComponentQualifier.Application
        }),
        branchLike: mockBranch()
      },
      MetricKey.projects
    )
  ).toMatchSnapshot('ref application component, projects');
  expect(
    shallowRender(
      {
        component: mockComponentMeasureEnhanced({
          refKey: 'project-key',
          qualifier: ComponentQualifier.Project
        }),
        branchLike: mockBranch()
      },
      MetricKey.projects
    )
  ).toMatchSnapshot('ref project component, projects');
  expect(
    shallowRender(
      {
        component: mockComponentMeasureEnhanced({
          refKey: 'app-key',
          qualifier: ComponentQualifier.Application
        }),
        branchLike: mockPullRequest()
      },
      MetricKey.alert_status
    )
  ).toMatchSnapshot('ref application component, alert_status metric');
  expect(
    shallowRender(
      {
        component: mockComponentMeasureEnhanced({
          refKey: 'vw-key',
          qualifier: ComponentQualifier.Portfolio
        }),
        branchLike: mockPullRequest()
      },
      MetricKey.alert_status
    )
  ).toMatchSnapshot('ref portfolio component, alert_status metric');
  expect(
    shallowRender({
      component: mockComponentMeasureEnhanced({
        key: 'svw-bar',
        qualifier: ComponentQualifier.SubPortfolio
      })
    })
  ).toMatchSnapshot('sub-portfolio component');
});

function shallowRender(overrides: Partial<ComponentCellProps> = {}, metricKey = MetricKey.bugs) {
  const metric = mockMetric({ key: metricKey });
  const component = enhanceComponent(
    mockComponentMeasure(true, {
      measures: [{ metric: metric.key, value: '1', bestValue: false }]
    }),
    metric,
    { [metric.key]: metric }
  );

  return shallow<ComponentCellProps>(
    <ComponentCell
      component={component}
      metric={metric}
      rootComponent={mockComponentMeasure(false)}
      view="list"
      {...overrides}
    />
  );
}
