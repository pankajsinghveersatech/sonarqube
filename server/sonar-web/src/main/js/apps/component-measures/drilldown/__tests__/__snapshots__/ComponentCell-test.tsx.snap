// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo:src/index.tsx"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo:src/index.tsx",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo:src/index.tsx"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="FIL"
        />
        <span
          className="note"
        >
          src
          /
        </span>
        <span>
          index.tsx
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: ref application component, alert_status metric 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/dashboard",
          "query": Object {
            "branch": undefined,
            "id": "app-key",
          },
        }
      }
    >
      <span
        className="big-spacer-right"
      >
        <LinkIcon />
      </span>
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="APP"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: ref application component, projects 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "branch": "branch-6.7",
            "id": "app-key",
            "metric": "projects",
            "view": "list",
          },
        }
      }
    >
      <span
        className="big-spacer-right"
      >
        <LinkIcon />
      </span>
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="APP"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: ref portfolio component, alert_status metric 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "vw-key",
            "metric": "alert_status",
            "pullRequest": "1001",
            "view": "list",
          },
        }
      }
    >
      <span
        className="big-spacer-right"
      >
        <LinkIcon />
      </span>
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="VW"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: ref project component 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "project-key",
            "metric": "bugs",
            "view": "list",
          },
        }
      }
    >
      <span
        className="big-spacer-right"
      >
        <LinkIcon />
      </span>
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: ref project component, projects 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/dashboard",
          "query": Object {
            "branch": "branch-6.7",
            "id": "project-key",
          },
        }
      }
    >
      <span
        className="big-spacer-right"
      >
        <LinkIcon />
      </span>
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: ref project component, releasability metric 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/dashboard",
          "query": Object {
            "branch": "branch-6.7",
            "id": "project-key",
          },
        }
      }
    >
      <span
        className="big-spacer-right"
      >
        <LinkIcon />
      </span>
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: root component is application, component has branch 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="TRK"
        />
        <span>
          Foo
        </span>
        <BranchIcon
          className="spacer-left little-spacer-right"
        />
        <span
          className="note"
        >
          develop
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: root component is application, component is on main branch 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-foo:src/index.tsx"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "foo:src/index.tsx",
            "view": "list",
          },
        }
      }
    >
      <span
        title="foo:src/index.tsx"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="FIL"
        />
        <span
          className="note"
        >
          src
          /
        </span>
        <span>
          index.tsx
        </span>
        <span
          className="spacer-left badge"
        >
          branches.main_branch
        </span>
      </span>
    </Link>
  </div>
</td>
`;

exports[`should render correctly: sub-portfolio component 1`] = `
<td
  className="measure-details-component-cell"
>
  <div
    className="text-ellipsis"
  >
    <Link
      className="link-no-underline"
      id="component-measures-component-link-svw-bar"
      onlyActiveOnIndex={false}
      style={Object {}}
      to={
        Object {
          "pathname": "/component_measures",
          "query": Object {
            "id": "foo",
            "metric": "bugs",
            "selected": "svw-bar",
            "view": "list",
          },
        }
      }
    >
      <span
        title="svw-bar"
      >
        <QualifierIcon
          className="little-spacer-right"
          qualifier="SVW"
        />
        <span>
          Foo
        </span>
      </span>
    </Link>
  </div>
</td>
`;
